File: scripts/gate-evaluator.js
# Creating the Node.js gate-evaluator script file and a README that explains usage.
script = r'''
#!/usr/bin/env node
/**
 * gate-evaluator.js
 *
 * Evaluator for cloud-native gates described in GATE_SPECIFICATIONS.md
 *
 * Usage:
 *   node gate-evaluator.js --snyk <snyk-results-dir> --sonar <sonar-results-dir> --thresholds <thresholds.json> --output <gate-result.json>
 *
 * The script:
 *  - Loads base thresholds (embedded)
 *  - Loads optional thresholds.json from repo (overrides, must be stricter)
 *  - Reads Snyk JSON results from provided directory (searches common filenames)
 *  - Reads Sonar results from provided directory (searches common filenames)
 *  - Evaluates gates: gatr-01, gatr-02, gatr-03, gatr-07, gatr-08, gatr-09, gatr-10, gatr-14
 *  - Writes gating result JSON and exits with code 0 (pass/warn) or 1 (fail enforcing)
 *
 * Notes:
 *  - This script tries to be defensive: if required data isn't present it will WARN rather than FAIL for NON_ENFORCING gates.
 *  - ENFORCING gates will FAIL when conditions matched unless a Jira exception file is provided (not implemented here).
 *
 * Output format: {
 *   final_decision: "PASS" | "WARN" | "FAIL" | "PASS_WITH_EXCEPTION",
 *   gates: [ { id, status, details... } ]
 * }
 */
const fs = require('fs');
const path = require('path');

function readJsonIfExists(p) {
  try {
    if (!p) return null;
    if (!fs.existsSync(p)) return null;
    const txt = fs.readFileSync(p, 'utf8');
    return JSON.parse(txt);
  } catch (e) {
    console.error('Failed to read/parse JSON', p, e.message);
    return null;
  }
}

function findFileInDir(dir, patterns) {
  if (!dir) return null;
  try {
    const files = fs.readdirSync(dir);
    for (const p of patterns) {
      const found = files.find(f => f.toLowerCase() === p.toLowerCase());
      if (found) return path.join(dir, found);
    }
    // try contains patterns
    for (const f of files) {
      for (const pat of patterns) {
        if (f.toLowerCase().includes(pat.toLowerCase())) return path.join(dir, f);
      }
    }
    return null;
  } catch (e) {
    return null;
  }
}

function loadSnykResults(dir) {
  // Common filenames
  const candidates = ['snyk-results.json', 'snyk-output.json', 'results.json', 'security-results.json'];
  const file = findFileInDir(dir, candidates);
  if (!file) return null;
  return readJsonIfExists(file);
}

function loadSonarResults(dir) {
  // Sonar might produce project_status or a custom json. Try common names.
  const candidates = ['sonar-report.json', 'sonar-results.json', 'project_status.json', 'sonar-quality.json', 'scan-report.json'];
  const file = findFileInDir(dir, candidates);
  if (!file) return null;
  return readJsonIfExists(file);
}

function metricRatingToNumber(r) {
  if (!r) return 99;
  const map = { 'A': 1, 'B': 2, 'C': 3, 'D': 4, 'E': 5 };
  return map[r] || 99;
}

function loadBaseThresholds() {
  return {
    snyk: { critical: 0, high: 5, medium: 20 },
    sonarqube: {
      coverage: 80,
      bugs: 5,
      vulnerabilities: 0,
      code_smells: 50,
      security_rating: 'A',
      reliability_rating: 'A',
      maintainability_rating: 'B',
      tech_debt_minutes: 480,
      express_lane: {
        coverage_threshold: 80,
        test_success_threshold: 80,
        max_security_rating: 'B',
        max_reliability_rating: 'C'
      }
    },
    approved_sonar_params: ['sonar.coverage.exclusions','sonar.cpd.exclusions']
  };
}

function mergeThresholds(base, overrides) {
  if (!overrides) return base;
  const merged = JSON.parse(JSON.stringify(base));
  if (overrides.snyk) {
    merged.snyk = Object.assign({}, merged.snyk, overrides.snyk);
  }
  if (overrides.sonarqube) {
    merged.sonarqube = Object.assign({}, merged.sonarqube, overrides.sonarqube);
  }
  return merged;
}

function evaluateSnykGates(snykJson, thresholds) {
  // Expect snykJson.vulnerabilities array
  const result = [];
  const vulns = (snykJson && snykJson.vulnerabilities) ? snykJson.vulnerabilities : [];
  const bySeverity = { critical:0, high:0, medium:0, low:0 };
  if (Array.isArray(vulns)) {
    for (const v of vulns) {
      const sev = (v.severity || v.issueSeverity || '').toLowerCase();
      if (bySeverity[sev] !== undefined) bySeverity[sev]++;
      else if (v.cvssScore) {
        const cv = Number(v.cvssScore || v.cvss || 0);
        if (cv >= 9) bySeverity.critical++;
        else if (cv >=7) bySeverity.high++;
        else if (cv >=4) bySeverity.medium++;
      }
    }
  }

  // gatr-03 Critical (non-enforcing but zero tolerance)
  if (bySeverity.critical > 0) {
    result.push({ id: 'gatr-03', status: 'WARN', critical_vulnerabilities: bySeverity.critical, threshold: thresholds.snyk.critical, message: 'Critical vulnerabilities detected - immediate attention required' });
  } else {
    result.push({ id: 'gatr-03', status: 'PASS', critical_vulnerabilities: 0, threshold: thresholds.snyk.critical });
  }

  // gatr-01 High
  if (bySeverity.high > thresholds.snyk.high) {
    result.push({ id: 'gatr-01', status: 'WARN', high_vulnerabilities: bySeverity.high, threshold: thresholds.snyk.high, message: 'High vulnerabilities exceed threshold' });
  } else {
    result.push({ id: 'gatr-01', status: 'PASS', high_vulnerabilities: bySeverity.high, threshold: thresholds.snyk.high });
  }

  // gatr-02 Medium
  if (bySeverity.medium > thresholds.snyk.medium) {
    result.push({ id: 'gatr-02', status: 'WARN', medium_vulnerabilities: bySeverity.medium, threshold: thresholds.snyk.medium, message: 'Medium vulnerabilities exceed threshold' });
  } else {
    result.push({ id: 'gatr-02', status: 'PASS', medium_vulnerabilities: bySeverity.medium, threshold: thresholds.snyk.medium });
  }

  return result;
}

function evaluateSonarGates(sonarJson, thresholds, options) {
  // sonarJson may be SonarCloud project_status or a custom metrics blob.
  const result = [];
  const sonarMetrics = (sonarJson && sonarJson.projectStatus) ? sonarJson.projectStatus : sonarJson;

  // Extract key metrics defensively
  let coverage = null, bugs = null, vulnerabilities = null, code_smells = null, ratings = {};
  // If Sonar's project_status with conditions:
  if (sonarMetrics && sonarMetrics.conditions && Array.isArray(sonarMetrics.conditions)) {
    for (const c of sonarMetrics.conditions) {
      if (c.metricKey === 'coverage') coverage = Number(c.actualValue);
      if (c.metricKey === 'new_coverage') coverage = Number(c.actualValue);
      if (c.metricKey === 'bugs') bugs = Number(c.actualValue);
      if (c.metricKey === 'vulnerabilities') vulnerabilities = Number(c.actualValue);
      if (c.metricKey === 'code_smells') code_smells = Number(c.actualValue);
    }
    if (sonarMetrics.status) {
      ratings.quality_gate_status = sonarMetrics.status;
    }
  } else if (sonarJson && typeof sonarJson === 'object') {
    // Try many possible shapes: metrics, measures, etc.
    if (sonarJson.metrics) {
      coverage = sonarJson.metrics.coverage || coverage;
      bugs = sonarJson.metrics.bugs || bugs;
      vulnerabilities = sonarJson.metrics.vulnerabilities || vulnerabilities;
      code_smells = sonarJson.metrics.code_smells || code_smells;
      if (sonarJson.metrics.ratings) ratings = sonarJson.metrics.ratings;
    }
    // fallback for the sample shape used earlier
    if (sonarJson.metrics && sonarJson.metrics.coverage !== undefined) coverage = sonarJson.metrics.coverage;
    if (sonarJson.qualityGate && sonarJson.qualityGate.status) ratings.quality_gate_status = sonarJson.qualityGate.status;
    if (sonarJson.metrics && sonarJson.metrics.ratings) ratings = sonarJson.metrics.ratings;
  }

  // gatr-07 Developer thresholds (NON_ENFORCING)
  const g07_issues = [];
  if (coverage !== null && coverage < thresholds.sonarqube.coverage) {
    g07_issues.push(`Coverage ${coverage}% < ${thresholds.sonarqube.coverage}%`);
  }
  if (bugs !== null && bugs > thresholds.sonarqube.bugs) {
    g07_issues.push(`Bugs ${bugs} > ${thresholds.sonarqube.bugs}`);
  }
  if (vulnerabilities !== null && vulnerabilities > thresholds.sonarqube.vulnerabilities) {
    g07_issues.push(`Vulnerabilities ${vulnerabilities} > ${thresholds.sonarqube.vulnerabilities}`);
  }
  if (code_smells !== null && code_smells > thresholds.sonarqube.code_smells) {
    g07_issues.push(`Code smells ${code_smells} > ${thresholds.sonarqube.code_smells}`);
  }
  if (g07_issues.length > 0) {
    result.push({ id: 'gatr-07', status: 'WARN', issues: g07_issues });
  } else {
    result.push({ id: 'gatr-07', status: 'PASS' });
  }

  // gatr-08 Code Quality (ENFORCING)
  // Trigger FAIL if quality gate status is ERROR or blocker issues > 0
  let g08_status = 'PASS';
  const g08_details = {};
  const qStatus = (ratings.quality_gate_status || '').toUpperCase();
  if (qStatus === 'ERROR' || qStatus === 'FAIL' ) {
    g08_status = 'FAIL';
    g08_details.message = 'Sonar Quality Gate status: ' + qStatus;
  }
  // also check for blocker issues if provided (some reports list blocker_issues)
  if (sonarJson && sonarJson.projectStatus && Array.isArray(sonarJson.projectStatus.conditions)) {
    const blockers = sonarJson.projectStatus.conditions.filter(c => (c.metricKey||'').includes('blocker') && c.status === 'ERROR');
    if (blockers.length > 0) {
      g08_status = 'FAIL';
      g08_details.blockers = blockers;
      g08_details.message = 'Blocker issues detected';
    }
  }
  result.push(Object.assign({ id: 'gatr-08', status: g08_status }, g08_details));

  // gatr-09 Approved Sonar Parameters (ENFORCING)
  // We will try to read a sonar-params file if provided in options.paramsFile
  const usedParams = options.sonarParams || [];
  const disallowed = usedParams.filter(p => !thresholds.approved_sonar_params.includes(p));
  if (disallowed.length > 0) {
    result.push({ id: 'gatr-09', status: 'FAIL', disallowed_parameters: disallowed, allowed_parameters: thresholds.approved_sonar_params });
  } else {
    result.push({ id: 'gatr-09', status: 'PASS', parameters_used: usedParams });
  }

  // gatr-10 Express Lane (NON_ENFORCING)
  const expressIssues = [];
  if (coverage !== null && coverage < thresholds.sonarqube.express_lane.coverage_threshold) {
    expressIssues.push(`Coverage ${coverage}% < ${thresholds.sonarqube.express_lane.coverage_threshold}%`);
  }
  // test_success_density may be present as metrics.test_success_density
  const test_success_density = sonarJson && sonarJson.metrics && sonarJson.metrics.test_success_density;
  if (test_success_density !== undefined && test_success_density < thresholds.sonarqube.express_lane.test_success_threshold) {
    expressIssues.push(`Test success density ${test_success_density}% < ${thresholds.sonarqube.express_lane.test_success_threshold}%`);
  }
  // ratings checks - try to read from ratings object
  if (ratings.security && metricRatingToNumber(ratings.security) > metricRatingToNumber(thresholds.sonarqube.express_lane.max_security_rating)) {
    expressIssues.push(`Security rating ${ratings.security} worse than allowed ${thresholds.sonarqube.express_lane.max_security_rating}`);
  }
  if (ratings.reliability && metricRatingToNumber(ratings.reliability) > metricRatingToNumber(thresholds.sonarqube.express_lane.max_reliability_rating)) {
    expressIssues.push(`Reliability rating ${ratings.reliability} worse than allowed ${thresholds.sonarqube.express_lane.max_reliability_rating}`);
  }
  if (expressIssues.length > 0) {
    result.push({ id: 'gatr-10', status: 'WARN', warnings: expressIssues });
  } else {
    result.push({ id: 'gatr-10', status: 'PASS' });
  }

  return result;
}

function evaluateBranchGate(env, options) {
  // gatr-14: ENFORCING
  const branchRef = env.GITHUB_REF || env.GITHUB_BRANCH || '';
  const targetEnv = env.TARGET_ENVIRONMENT || options.target || '';
  const allowedPatterns = [/^refs\/heads\/main$/, /^refs\/heads\/release\/.+$/];

  const isAllowed = allowedPatterns.some(p => p.test(branchRef));
  if ((targetEnv === 'PROD' || targetEnv === 'UAT') && !isAllowed) {
    return { id: 'gatr-14', status: 'FAIL', branch: branchRef, environment: targetEnv, message: 'Only main or release/* branches can deploy to UAT/PROD' };
  }
  return { id: 'gatr-14', status: 'PASS', branch: branchRef, environment: targetEnv };
}

function decideFinal(gates) {
  // If any ENFORCING gate is FAIL -> overall FAIL
  // ENFORCING: gatr-08, gatr-09, gatr-14
  const enforcingIds = ['gatr-08','gatr-09','gatr-14'];
  let final = 'PASS';
  for (const g of gates) {
    if (enforcingIds.includes(g.id) && g.status === 'FAIL') {
      final = 'FAIL';
      return final;
    }
    if (g.status === 'WARN' && final !== 'FAIL') {
      final = 'WARN';
    }
  }
  return final;
}

// ---------------------
// CLI parsing
// ---------------------
const argv = require('minimist')(process.argv.slice(2), {
  string: ['snyk','sonar','thresholds','output','sonar-params-file','target'],
  alias: { s:'snyk', q:'sonar', t:'thresholds', o:'output' },
  default: {}
});

const snykDir = argv.snyk;
const sonarDir = argv.sonar;
const thresholdsFile = argv.thresholds;
const outFile = argv.output || 'gating/gate-result.json';
const sonarParamsFile = argv['sonar-params-file'];
const target = argv.target || process.env.TARGET_ENVIRONMENT;

const baseThresholds = loadBaseThresholds();
const repoOverrides = readJsonIfExists(thresholdsFile);
const thresholds = mergeThresholds(baseThresholds, repoOverrides);

const snykJson = loadSnykResults(snykDir);
const sonarJson = loadSonarResults(sonarDir);

let sonarParams = [];
if (sonarParamsFile && fs.existsSync(sonarParamsFile)) {
  try {
    sonarParams = JSON.parse(fs.readFileSync(sonarParamsFile,'utf8'));
    if (!Array.isArray(sonarParams)) sonarParams = [];
  } catch(e) { sonarParams = []; }
}

// Evaluate
const gateResults = [];
gateResults.push(...evaluateSnykGates(snykJson, thresholds));
gateResults.push(...evaluateSonarGates(sonarJson, thresholds, { sonarParams, target }));
gateResults.push(evaluateBranchGate(process.env, { target }));

const finalDecision = decideFinal(gateResults);

// Ensure output dir exists
try {
  const outDir = path.dirname(outFile);
  if (!fs.existsSync(outDir)) fs.mkdirSync(outDir, { recursive: true });
  fs.writeFileSync(outFile, JSON.stringify({ final_decision: finalDecision, gates: gateResults }, null, 2), 'utf8');
  console.log('Gate result written to', outFile);
} catch (e) {
  console.error('Failed to write output', e.message);
  process.exit(2);
}

// Exit codes: 0 for PASS/WARN/PASS_WITH_EXCEPTION, 1 for FAIL
if (finalDecision === 'FAIL') {
  console.error('Gate evaluation result: FAIL');
  process.exit(1);
} else {
  console.log('Gate evaluation result:', finalDecision);
  process.exit(0);
}
'''.strip()

readme = r'''
Gate Evaluator - README
-----------------------

Files created:
 - /mnt/data/gate-evaluator.js

How to use in GitHub Actions:
 - Place this script in repository under scripts/gate-evaluator.js
 - Make sure Node.js is available in runner (ubuntu-latest has node)
 - Example step:

    - name: Run Gate Evaluator
      run: |
        chmod +x scripts/gate-evaluator.js
        node scripts/gate-evaluator.js \
          --snyk "gating/security/unzipped" \
          --sonar "gating/quality/unzipped" \
          --thresholds "thresholds.json" \
          --sonar-params-file "gating/quality/sonar-params.json" \
          --output "gating/gate-result.json"

Notes:
 - The script expects Snyk JSON with a 'vulnerabilities' array and Sonar results in common names (sonar-report.json, sonar-results.json, project_status.json).
 - Adjust paths as needed.
 - If you want, I can also generate a Python version or a more extensive test harness that runs locally against your uploaded zip files.
'''

# Write files to /mnt/data
import os
os.makedirs('/mnt/data/scripts', exist_ok=True)
with open('/mnt/data/scripts/gate-evaluator.js', 'w', encoding='utf8') as f:
    f.write(script)
with open('/mnt/data/scripts/GATE_EVALUATOR_README.md', 'w', encoding='utf8') as f:
    f.write(readme)

# Provide download link to user
print("Created /mnt/data/scripts/gate-evaluator.js and README")
print("Download the script: sandbox:/mnt/data/scripts/gate-evaluator.js")
print("Download the README: sandbox:/mnt/data/scripts/GATE_EVALUATOR_README.md")

File: scripts/gate_evaluator.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
gate_evaluator.py
------------------
Evaluador Python para resultados de Snyk + SonarQube, basado en las reglas
definidas en GATE_SPECIFICATIONS.md.

Este script:
    - Carga resultados de Snyk (SCA y SAST)
    - Carga resultados de SonarQube
    - Aplica umbrales de calidad predefinidos o personalizados
    - Evalúa múltiples reglas GATR-XX
    - Genera un archivo JSON final con el resultado
    - Retorna un código de salida para enforcement

Uso:
    python gate_evaluator.py \
        --snyk results/security \
        --sonar results/quality \
        --thresholds gating/thresholds.json \
        --params gating/sonar-params.json \
        --output gating/gate-result.json \
        --target PROD \
        --ref refs/heads/release/1.0.0

Exit Codes:
    0 → PASS / WARN / PASS_WITH_EXCEPTION
    1 → FAIL
"""

import os
import json
import argparse
import requests
from datetime import datetime
from pathlib import Path


# ======================================================================
# Jira Exception Checker
# ======================================================================

def jira_check_exception(gate_id, app_id):
    """
    Consulta Jira para verificar si existe una excepción aprobada
    para un gate específico.
    """
    JIRA_URL = os.getenv("JIRA_URL")
    JIRA_USER = os.getenv("JIRA_USER")
    JIRA_TOKEN = os.getenv("JIRA_API_TOKEN")
    JIRA_PROJECT = os.getenv("JIRA_PROJECT", "GATES")

    if not all([JIRA_URL, JIRA_USER, JIRA_TOKEN]):
        print("⚠ Jira integration not configured. Skipping.")
        return {"approved": False}

    today = datetime.utcnow().strftime("%Y-%m-%d")

    jql = f'''
        project = {JIRA_PROJECT} AND
        cf_gate_id = "{gate_id}" AND
        cf_application_id = "{app_id}" AND
        cf_exception_approval_status = "DECISION MADE" AND
        cf_exception_approval_decision = "Approved" AND
        cf_exception_expiry_date >= "{today}"
    '''

    url = f"{JIRA_URL}/rest/api/2/search"

    try:
        response = requests.get(
            url,
            params={"jql": jql},
            auth=(JIRA_USER, JIRA_TOKEN),
            headers={"Content-Type": "application/json"}
        )
    except Exception as e:
        print(f"Error contacting Jira: {e}")
        return {"approved": False}

    if response.status_code != 200:
        print(f"Jira query failed: HTTP {response.status_code}")
        return {"approved": False}

    data = response.json()

    if data.get("total", 0) > 0:
        issue = data["issues"][0]
        return {
            "approved": True,
            "issue_key": issue["key"],
            "expiry": issue["fields"].get("cf_exception_expiry_date")
        }

    return {"approved": False}

# ======================================================================
# Utility Helpers
# ======================================================================

def read_json(path):
    if not path or not os.path.exists(path):
        return None
    try:
        with open(path, "r", encoding="utf-8") as f:
            return json.load(f)
    except:
        return None


def find_file(directory, candidates):
    if not directory or not os.path.exists(directory):
        return None

    files = os.listdir(directory)

    # exact match
    for c in candidates:
        if c.lower() in [f.lower() for f in files]:
            return os.path.join(directory, c)

    # partial match
    for f in files:
        for c in candidates:
            if c.lower() in f.lower():
                return os.path.join(directory, f)

    return None


# ======================================================================
# Load Snyk & SonarQube Results
# ======================================================================

def load_snyk_results(directory):
    candidates = [
        "snyk-results.json", "snyk-output.json", "results.json",
        "security-results.json", "snyk-report.json"
    ]
    f = find_file(directory, candidates)
    return read_json(f)


def load_sonar_results(directory):
    candidates = [
        "sonar-report.json", "sonar-results.json", "project_status.json",
        "sonar-quality.json", "scan-report.json"
    ]
    f = find_file(directory, candidates)
    return read_json(f)


# ======================================================================
# Thresholds
# ======================================================================

def default_thresholds():
    return {
        "snyk": {"critical": 0, "high": 5, "medium": 20},
        "sonarqube": {
            "coverage": 80,
            "bugs": 5,
            "vulnerabilities": 0,
            "code_smells": 50,
            "security_rating": "A",
            "reliability_rating": "A",
            "maintainability_rating": "B",
            "tech_debt_minutes": 480,
            "express_lane": {
                "coverage_threshold": 80,
                "test_success_threshold": 80,
                "max_security_rating": "B",
                "max_reliability_rating": "C"
            }
        },
        "approved_sonar_params": ["sonar.coverage.exclusions", "sonar.cpd.exclusions"]
    }


def merge_thresholds(base, overrides):
    if not overrides:
        return base
    merged = json.loads(json.dumps(base))
    if "snyk" in overrides:
        merged["snyk"].update(overrides["snyk"])
    if "sonarqube" in overrides:
        merged["sonarqube"].update(overrides["sonarqube"])
    return merged


# ======================================================================
# Evaluation: Snyk
# ======================================================================

def evaluate_snyk(snyk_json, t):
    results = []
    vulns = snyk_json.get("vulnerabilities", []) if snyk_json else []

    sev = {"critical": 0, "high": 0, "medium": 0}

    for v in vulns:
        s = v.get("severity", "").lower()
        if s in sev:
            sev[s] += 1

    # GATR-03 Critical
    results.append({
        "id": "gatr-03",
        "status": "PASS" if sev["critical"] == t["snyk"]["critical"] else "WARN",
        "count": sev["critical"],
        "threshold": t["snyk"]["critical"]
    })

    # GATR-01 High
    results.append({
        "id": "gatr-01",
        "status": "PASS" if sev["high"] <= t["snyk"]["high"] else "WARN",
        "count": sev["high"],
        "threshold": t["snyk"]["high"]
    })

    # GATR-02 Medium
    results.append({
        "id": "gatr-02",
        "status": "PASS" if sev["medium"] <= t["snyk"]["medium"] else "WARN",
        "count": sev["medium"],
        "threshold": t["snyk"]["medium"]
    })

    return results


# ======================================================================
# Evaluation: SonarQube
# ======================================================================

def evaluate_sonar(sonar_json, t, used_params):
    if not sonar_json:
        return [{"id": "gatr-07", "status": "WARN", "message": "Missing Sonar results"}]

    metrics = sonar_json.get("metrics", {})
    ratings = metrics.get("ratings", {})

    issues = []

    def fail_if(cond, msg):
        if cond:
            issues.append(msg)

    fail_if(metrics.get("coverage") < t["sonarqube"]["coverage"],
            f"Coverage {metrics.get('coverage')} < {t['sonarqube']['coverage']}")

    fail_if(metrics.get("bugs") > t["sonarqube"]["bugs"],
            f"Bugs exceed {t['sonarqube']['bugs']}")

    fail_if(metrics.get("vulnerabilities") > t["sonarqube"]["vulnerabilities"],
            f"Vulns exceed {t['sonarqube']['vulnerabilities']}")

    fail_if(metrics.get("code_smells") > t["sonarqube"]["code_smells"],
            f"Code smells exceed {t['sonarqube']['code_smells']}")

    fail_if(ratings.get("security") != t["sonarqube"]["security_rating"],
            "Security rating mismatch")

    fail_if(ratings.get("reliability") != t["sonarqube"]["reliability_rating"],
            "Reliability rating mismatch")

    fail_if(ratings.get("maintainability") != t["sonarqube"]["maintainability_rating"],
            "Maintainability rating mismatch")

    # GATR-07
    g07 = {
        "id": "gatr-07",
        "status": "PASS" if not issues else "WARN",
        "issues": issues
    }

    # GATR-08 Quality Gate
    q_status = sonar_json.get("quality_gate", {}).get("status", "OK")
    g08 = {
        "id": "gatr-08",
        "status": "FAIL" if q_status.upper() in ["ERROR", "FAIL"] else "PASS",
        "quality_gate_status": q_status
    }

    # GATR-09 Allowed parameters
    disallowed = [p for p in used_params if p not in t["approved_sonar_params"]]
    g09 = {
        "id": "gatr-09",
        "status": "FAIL" if disallowed else "PASS",
        "disallowed": disallowed
    }

    return [g07, g08, g09]


# ======================================================================
# GATR-14 Branch Validation
# ======================================================================

def evaluate_branch(ref, env):
    if ref.startswith("refs/heads/release/"):
        return {"id": "gatr-14", "status": "PASS", "branch": ref}

    if env in ("PROD", "UAT") and ref != "refs/heads/main":
        return {"id": "gatr-14", "status": "FAIL", "branch": ref,
                "message": "Only main or release/* allowed"}

    return {"id": "gatr-14", "status": "PASS", "branch": ref}


# ======================================================================
# GATR-10 Express Lane
# ======================================================================

def evaluate_express_lane(sonar_json, t):
    if not sonar_json:
        return {"id": "gatr-10", "status": "WARN", "message": "Missing Sonar"}

    metrics = sonar_json.get("metrics", {})
    ratings = metrics.get("ratings", {})
    new = metrics.get("new_ratings", {})

    e = t["sonarqube"]["express_lane"]
    issues = []

    if metrics.get("coverage") < e["coverage_threshold"]:
        issues.append("Coverage too low")

    if metrics.get("test_success_rate") < e["test_success_threshold"]:
        issues.append("Test success too low")

    if ratings.get("security") > e["max_security_rating"]:
        issues.append("Security rating too low")

    if ratings.get("reliability") > e["max_reliability_rating"]:
        issues.append("Reliability rating too low")

    if new.get("security") not in (None, "A"):
        issues.append("New-code security != A")

    if new.get("reliability") not in (None, "A"):
        issues.append("New-code reliability != A")

    return {
        "id": "gatr-10",
        "status": "PASS" if not issues else "WARN",
        "issues": issues
    }


# ======================================================================
# FINAL DECISION
# ======================================================================

def decide_final(gates):
    enforcing = {"gatr-08", "gatr-09", "gatr-14"}
    final = "PASS"
    for g in gates:
        if g["id"] in enforcing and g["status"] == "FAIL":
            return "FAIL_PENDING_EXCEPTION"
        if g["status"] == "WARN" and final == "PASS":
            final = "WARN"
    return final


# ======================================================================
# MAIN
# ======================================================================

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--snyk")
    parser.add_argument("--sonar")
    parser.add_argument("--thresholds")
    parser.add_argument("--params")
    parser.add_argument("--output", default="gating/gate-result.json")
    parser.add_argument("--target", default="DEV")
    parser.add_argument("--ref", default="refs/heads/main")
    args = parser.parse_args()

    # Cargar thresholds
    base = default_thresholds()
    overrides = read_json(args.thresholds)
    thresholds = merge_thresholds(base, overrides)

    # Cargar resultados
    snyk_json = load_snyk_results(args.snyk)
    sonar_json = load_sonar_results(args.sonar)
    sonar_params = read_json(args.params) or []

    # Ejecutar evaluaciones
    gates = []
    gates.extend(evaluate_snyk(snyk_json, thresholds))
    gates.append(evaluate_express_lane(sonar_json, thresholds))
    gates.extend(evaluate_sonar(sonar_json, thresholds, sonar_params))
    gates.append(evaluate_branch(args.ref, args.target))

    # Decisión sin excepciones
    final = decide_final(gates)

    # Si falla un gate ENFORCING → buscar excepción en Jira
    if final == "FAIL_PENDING_EXCEPTION":
        app_id = os.getenv("APP_ID", "unknown")
        print("Checking Jira for exception approvals...")

        failed = [
            g for g in gates
            if g["id"] in ("gatr-08", "gatr-09", "gatr-14") and g["status"] == "FAIL"
        ]

        exception_found = False

        for g in failed:
            r = jira_check_exception(g["id"], app_id)
            if r.get("approved"):
                g["status"] = "PASS_WITH_EXCEPTION"
                g["exception_issue"] = r["issue_key"]
                g["exception_expiry"] = r["expiry"]
                exception_found = True
                print(f"✔ Exception approved for {g['id']} ({r['issue_key']})")

        if exception_found:
            final = "PASS_WITH_EXCEPTION"
        else:
            final = "FAIL"

    # Crear carpeta
    out_dir = Path(args.output).parent
    out_dir.mkdir(parents=True, exist_ok=True)

    # Guardar salida
    with open(args.output, "w", encoding="utf-8") as f:
        json.dump({"final_decision": final, "gates": gates}, f, indent=2)

    print(f"Gate evaluation: {final}")
    print(f"Output written to: {args.output}")

    exit(1 if final == "FAIL" else 0)


if __name__ == "__main__":
    main()
